pkg/agent/openflow/pipeline.go

```diff
@@ -279,6 +306,18 @@ func (c *client) l2ForwardOutputFlow(category cookie.Category) binding.Flow {
 		Done()
 }
 
+// l2ForwardOutputFlow generates the flow that forward peer node traffic via gw0.
+// This flow supersedes default output flow because ovs by default auto-skips packets with output = input port.
+func (c *client) l2ForwardOutputToRemoteInPortOut(gwPort uint32, category cookie.Category) binding.Flow {
+	return c.pipeline[l2ForwardingOutTable].BuildFlow(priorityHigh).MatchProtocol(binding.ProtocolIP).
+		MatchRegRange(int(marksReg), portFoundMark, ofPortMarkRange).
+		MatchInPort(gwPort).MatchRegRange(int(portCacheReg), gwPort, ofPortRegRange).
```

Comment:
There might be two kinds of traffic sent to the OVS bridge: 1) local pod to remote pod 2) local Pod to Service, and then to the backend Pod. For 1), this logic is correct, for we want to use host routing/iptables to forward the packet to the remote Node. For 2), if the backend Pod is also on the same node, this flow might send the packet to a wrong target.

Reply:
This flows refers service traffic enters gw0 to LB and goes out again to a pod on a different node.
if pod backing a service is from local node, the load balanced pod-to-pod traffic will not hit this flow.

say backend pod is on the same node, service traffic will be

request: container(dstIP=serviceIP, srcIP=containerIP) ===> gw0(egress to host network, inport != outpor = gw0, this flow is not hit) ==> LB (dstIP=backEndlContainerIP, srcIP=serviceIP) ==> gw0(ingress from host network) ===> backend container via l2 flow 

Notice in this data flow the l2ForwardOutputToRemoteInPortOut flow was not hit?

Reply:
Got it, you also use the destinationed OpenFlow port(gw0) as one match condition in this flow, so it would not be hit by the packet targets at local containers. Thanks for explanation.